File: v1/src/main/python/transform.py
"""
Copyright 2020 Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import json
import sys

def transform(event):
  """ Return a Dict or List of Dict Objects.  Return None to discard """
  # Ensure Table Dest Exists

  return clean_transformed_event(event)

def clean_transformed_event(event):
  # TODO: Review logic to decide if cleaning all complex data is required
  for col in event:
    if type(event[col]) in [dict, list]:
      event[col] = json.dumps(event[col])

  return json.dumps(event)

def _handle_result(result):
  if isinstance(result, list):
    for event in result:
      if event:
        print(event)
  elif result:
    print(result)

if __name__ == '__main__':
  # TODO: How do we want to handle the case where there are no messages
  file_name = sys.argv[1]
  with open(file_name, "r") as data_file:
    for cnt, raw_data in enumerate(data_file):
      raw_events = "[%s]" % raw_data

      # TODO: What if line is not valid JSON, how are errors pushed to DLQ
      data = json.loads(raw_events)

      if isinstance(data, list):
        for event in data:
          _handle_result(transform(event))
      else:
        event = data
        _handle_result(transform(event))
  exit()

File: v1/src/test/resources/JavascriptTextTransformerTest/transform.js
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/**
 * A good transform function.
 * @param {string} inJson
 * @return {string} outJson
 */
function transform(inJson) {
  var obj = JSON.parse(inJson);
  obj.someProp = "someValue";
  return JSON.stringify(obj);
}

/**
 * A transform function which only accepts 42 as the answer to life.
 * @param {string} inJson
 * @return {string} outJson
 */
function transformWithFilter(inJson) {
  var obj = JSON.parse(inJson);
  // only output objects which have an answer to life of 42.
  if (!(obj.hasOwnProperty('answerToLife') && obj.answerToLife != 42)) {
    return JSON.stringify(obj);
  }
}
/**
 * A transform function which copies data to another attribute, with computation in the middle.
 * @param {string} inJson
 * @return {string} outJson
 */
function transformSlow(inData) {

  var row = inData.split(',');
  var obj = new Object();

  obj.name = row[0];

  sum = 0;
  for (let i = 0; i <= 10000; i++) {
    sum += i;
  }
  obj.model = row[1];
  obj.sum = sum;

  return JSON.stringify(obj);
}